### ISSUE TITLE
git-vogue checks workding dir files, not whas is actually commited

### ISSUE THREAD

**Issue by ehamberg** [anchor/git-vogue#88]
(This might be related to #82.)

If I add some debugging code to some module that would not pass `git-vogue`'s checks, but exclude that when building my commit (using `git add -p`), it seems that `git-vogue` still rejects my commit.

I really think `git-vogue` should only check what's actually being committed, not what my working dir happens to contain.

**Comment by christian-marie** [comment]
Can you provide a repository on which I can reproduce this? I've not noticed such behavior.

**Comment by ehamberg** [comment]
Sure. Her is a (slightly contrived) example:

    ❯ git clone git@github.com:ehamberg/git-vogue-test.git
    ❯ cd git-vogue-test
    ❯ git reset HEAD~1
    Unstaged changes after reset:
    M    Main.hs
    
    ❯ git vogue init
    Successfully installed hook

Okay, there are two changes at this point. One is the `import` lines that will fail the *stylish-haskell* check, and the other is the definition `f = succ`. We don't want to commit the import lines, only the definition of `f`:

    ❯ git add --patch
    diff --git a/Main.hs b/Main.hs
    index 9ed23e9..975bc09 100644
    --- a/Main.hs
    +++ b/Main.hs
    @@ -1,3 +1,8 @@
     module Main where
    
    +import Data.List
    +import Control.Monad
    +
     main = putStrLn "Hi."
    +
    +f = succ
    Stage this hunk [y,n,q,a,d,/,s,e,?]? s
    Split into 2 hunks.
    @@ -1,3 +1,6 @@
     module Main where
    
    +import Data.List
    +import Control.Monad
    +
     main = putStrLn "Hi."
    Stage this hunk [y,n,q,a,d,/,j,J,g,e,?]? n
    @@ -3 +6,3 @@
     main = putStrLn "Hi."
    +
    +f = succ
    Stage this hunk [y,n,q,a,d,/,K,g,e,?]? y

Okay. At this point we have staged the hunk

    +
    +f = succ

The `import` lines hunk is *not* staged, so git vogue should be happy. However, git-vogue looks at what's in the file in the working dir, not how the file will look like after the commit, so the commit fails:

    ❯ git ci -m 'XXX'
    cabal succeeded with:
    
    ghc-mod succeeded with:
    
    hlint succeeded with:
    
    stylish failed with:
      [-] Main.hs has differing style:
          1,2c1,2
          < import Data.List
          < import Control.Monad
          ---
          > import           Control.Monad
          > import           Data.List

These issues most often show up when playing around with adding e.g. the `Debug.Trace` import and adding some `trace` calls. Even when excluding these hunks, `git-vogue` will reject commits. Another example I've seen is adding `-fhpc` to the GHC flags in a cabal file to see test coverage. Again, even when not adding that change to a commit, `git-vogue` will reject the commit.

Of course, a work-around is to first commit and verify that the errors from `git-vogue` does not apply to what I'm going to commit and then use `git commit --no-verify`, but this is error-prone, and I'd really *like* to have `git-vogue` check what I'm committing. :-)

**Comment by thsutton** [comment]
Yeah, I expect this could be resolved by having `git-vogue` use the VCS' facilities to read files. I git we'll want to read from the index, not the file on disk. Probably using `git ls-files` and `git cat-file`.

**Comment by christian-marie** [comment]
Ah. I understand what you are talking about now and agree that @thsutton's approach is one way forward.

Alternately we could stash everything, run the check, then pop it? I don't want to make plugins aware of VCS if that's at all possible.

**Comment by christian-marie** [comment]
@tvh, @tranma, @thsutton, @oswynb, opinions?

**Comment by thsutton** [comment]
@christian-marie Is it feasible to make "read the contents of file X" part of the VCS implementation? So plugins can call a library function or even just, e.g., `git-vogue read-file Main.hs` to get the code to check.

Then `fix` will have to bail out if there are files that have both staged and unstaged changes.

**Comment by christian-marie** [comment]
This is complicated, and now plugins such as cabal-check that do not get given file contents will need to create temp dirs, re-create the required things, check in that dir, and hope that it doesn't require access to the current "build".

OTOH, this is indeed feasible, any cleaner approaches?

**Comment by expipiplus1** [comment]
A nicer workaround would be to use `git stash -k` to stash unstaged changes.

**Comment by christian-marie** [comment]
This just needs to be done.

**Comment by christian-marie** [comment]
I'm not sure what the type would be
